src/types.rs

```diff
@@ -1228,27 +1229,56 @@ impl SourceMapIndex {
                 }
             };
 
+            for (source, contents) in map.sources().zip(map.source_contents()) {
+                let src_id = builder.add_source(source);
+
+                if let Some(contents) = contents {
+                    builder.set_source_contents(src_id, Some(contents));
+                }
+            }
+
+            let mut src_id_map = FxHashMap::<u32, u32>::default();
+            let mut name_id_map = FxHashMap::<u32, u32>::default();
+
             for token in map.tokens() {
                 let dst_col = if token.get_dst_line() == 0 {
                     token.get_dst_col() + off_col
                 } else {
                     token.get_dst_col()
                 };
-                let raw = builder.add(
+
+                // Use u32 -> u32 map instead of using the hash map in SourceMapBuilder for better performance
+                let original_src_id = token.raw.src_id;
+                let src_id = if original_src_id == !0 {
+                    None
+                } else {
+                    Some(*src_id_map.entry(original_src_id).or_insert_with(|| {
```

Comment:
In case you want to pre-build the `src_id_map`, you could just `src_id_map.get(orginial_src_id).copied()` here.

I think it might make sense to pre-populate the `name_id_map` in a similar way as well.